R/tree-sequences.R

```diff
@@ -120,26 +138,31 @@ ts_load <- function(model, file = NULL,
   reticulate::source_python(file = system.file("pylib/pylib.py", package = "slendr"))
 
   attr(ts, "source") <- backend
+  attr(ts, "spatial") <- backend == "SLiM" && ts$metadata$SLiM$spatial_dimensionality != ""
 
   attr(ts, "model") <- model
-  attr(ts, "metadata") <- get_metadata(ts)
+  if (!is.null(model))
+    attr(ts, "metadata") <- get_slendr_metadata(ts)
 
   attr(ts, "recapitated") <- FALSE
   attr(ts, "simplified") <- FALSE
   attr(ts, "mutated") <- FALSE
 
   class(ts) <- c("slendr_ts", class(ts))
 
-  attr(ts, "nodes") <- get_ts_nodes(ts)
-  attr(ts, "edges") <- get_ts_edges(ts)
+  attr(ts, "raw_nodes") <- get_ts_raw_nodes(ts)
+  attr(ts, "raw_edges") <- get_ts_raw_edges(ts)
+  attr(ts, "raw_individuals") <- get_ts_raw_individuals(ts)
 
   if (backend == "SLiM") {
-    attr(ts, "individuals") <- get_ts_individuals(ts) %>% dplyr::mutate(sampled = remembered)
-    attr(ts, "data") <- get_slim_table_data(ts, model, spatial)
+    remembered <- attr(ts, "raw_individuals")$remembered
+    if (!any(remembered)) remembered <- attr(ts, "raw_individuals")$alive
+    attr(ts, "raw_individuals")$sampled <- remembered
+    attr(ts, "nodes") <- get_slim_table_data(ts, model, spatial)
   } else {
-    attr(ts, "individuals") <- get_ts_individuals(ts)
-    attr(ts, "data") <- get_msprime_table_data(ts, model)
-    }
+    attr(ts, "raw_individuals")$sampled <- TRUE
```

Comment:
Isn't the right way to do this to use the individuals' node flags?

Reply:
Yes, the individuals' node flags are always present and accessible to the user. However, I discovered that a tree sequence in which every (or most) individuals are "remembered" breaks lots of my code.

For instance, attempting to extract a single tree from such tree sequence and converting it to a standard phylogenetic tree object (all that I evaluated assume only tips are samples) broke because even if a user specified only a subset of "remembered" individuals to be used for simplification (out of every other node which was also "remembered"), other internal nodes in the simplified tree sequence were also remembered... and slendr had no way of knowing which of them are those that the user is interested in (i.e. the subset they provided to guide the simplification process) and which are "not interesting" -- internal nodes NOT provided for simplification -- but still carrying the remembered flag (because every single individual has been remembered).

This is why I added a separate tag that I'm carrying along in the slendr-munged tables called `sampled`. When a tree sequence is first loaded, every "remembered" individual assumed to be "sampled" (because the only thing this can happen in slendr is by being marked as sampled from the R interface), but when that tree sequence is simplified down to a subset of (potentially very many) remembered nodes, slendr keeps track of which nodes are interesting to the user (called "samples") and which are not (everything that's not `sampled == TRUE`).

Reply:
I do admit that it's been all motivated by being able to use these "edge-case" tree sequences (i.e. everyone who ever lived is explicitly remembered in the tree sequence) with the standard phylogenetics packages I've looked at.

Maybe there's a better solution that I'm not seeing right now. But

a) all the pyslim flags are still present and available (I'm not reinventing them here)
b) this is only needed in a couple of places in the whole codebase, so if a better way to do this comes up it's not like the whole thing will have to be re-written from scratch (the "sampled" column became a thing only 2 months ago after  months of not being needed... until I hit that "everyone is remembered" fringe case).

Reply:
Gee, this is confusing. So, what's the problem with 'everyone remembered' tree sequences? What breaks when you try to get a tree?

Reply:
Hmm, OK, I have to take that one statement back. My brain got confused there for a minute, mixing stuff up with the topic of [discussion](https://github.com/tskit-dev/tskit/discussions/2144) I originally started in the tskit forum on the +/- same topic. The workaround I implemented for getting the ape R phylo tree in the end doesn't have a problem with the "everyone is remembered" tree sequences anymore.

---

[... fast forward through Saturday + Sunday refactoring frenzy ...]

---

... also, as I was just about to send a justification for why I want/need that extra column "sampled", I re-read the documentation of tskit & pyslim yet again and discovered that the reason I wanted this extra column -- for the many situations in which a user wants to know which are **focal** individuals/nodes, not just remembered ones -- is actually captured by `TreeSequence.samples()`? Unless I'm missing something obvious, this is exactly what I basically re-implemented above.

I spent the whole weekend removing this, putting this back in, then renaming this column to `focal` because it fitted better to the semantics I had in mind for this... then I noticed that I'm probably reinventing the wheel (not realizing that remembered-retained-alive are not the only things that are tracked, but that a node can be remembered yet not be a sample).

**Just to make sure before I spend even more hours in the tskit rabbit hole: if a user simulates a tree sequence in which every single individual/node is remembered (let's say 10000 nodes are remembered out of 10000 nodes in the tree sequence total), then only simplifies down to a subset of 20 nodes, `ts.samples()` will only get them IDs of those 20 nodes, right? Every other node in that simplified tree sequence will be still flagged as `remembered` (probably less than 10000 we started with, but still likely quite many nodes), right?** Maybe @bhaller could help me here too? 

So, just to explain the code that probably did not make much sense to you @petrelharp (that `$sampled` column business), I think I was reinventing exactly what `ts.samples()` does. I wanted to give the user access to information about which nodes (out of possible many remembered) are the "focal nodes" they simplified to, and the chance to ignore the rest (even those they also explicitly remembered at first).

I'm astonished that even after 9 months of working with this I "discover" this "new feature". 🤦 I guess my excuse is that I have never used tree sequences outside of the slendr context...

Reply:
If the bold paragraph in the previous message is correct, then the `sampled` column would be simply populated by `TRUE` for whatever `node_id` is among `ts.samples()` and `FALSE` for any other node.

That is, unless I am -- for the 100th time -- missing something that's probably obvious to everyone else.

Reply:
> ...Maybe @bhaller could help me here too?

I'm afraid I cannot; you are already way beyond my level of expertise with tskit!  I mostly (but not entirely) understand how tree-sequence recording inside SLiM works.  My knowledge of other tskit things is fairly minimal.

> That is, unless I am -- for the 100th time -- missing something that's probably obvious to everyone else.

Don't beat yourself up, @bodkan – you're doing great and this stuff is amazing.  Learning to use a complex library like tskit always takes some iteration.

Reply:
Right, ok, let's see:
1. everyone who is `remembered` will have their nodes marked as samples in the output tree sequence
2. everyone who is `retained` but not remembered will *not* be marked as samples
3. `simplify(x)` *only* leaves those nodes in `x` marked as samples but does not touch individual flags
4. there is no currently convenient way to "get individuals whose nodes are samples" although there **definitely should be**, we just haven't figured out the right API and done it yet

(Is this answering your question?  Maybe we should have some discussion on things like this on slack or even zoom - I don't always notice the github things right away.)